app/models/mixins/authentication_mixin.rb

```diff
@@ -168,21 +168,13 @@ def update_authentication(data, options = {})
       end
 
       # Update or create
-      if cred.nil?
-        if self.kind_of?(ManageIQ::Providers::Openstack::InfraManager) && value[:auth_key]
-          # TODO(lsmola) investigate why build throws an exception, that it needs to be subclass of AuthUseridPassword
-          cred = ManageIQ::Providers::Openstack::InfraManager::AuthKeyPair.new(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                               :resource_id => id, :resource_type => "ExtManagementSystem")
-          authentications << cred
-        elsif value[:auth_key]
-          cred = AuthToken.new(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                               :resource_id => id, :resource_type => "ExtManagementSystem")
-          authentications << cred
-        else
-          cred = authentications.build(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                            :type => "AuthUseridPassword")
-        end
-      end
+      cred ||= authentication_class(value.merge(:authtype => type.to_s)).new(
+        :name          => "#{self.class.name} #{name}",
+        :authtype      => type.to_s,
+        :resource_id   => id,
+        :resource_type => 'ExtManagementSystem'
```

Comment:
Not for this PR but I wonder why this wasn't just `:resource => self`?

Reply:
Yeah, not in this PR please :laughing: